src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -712,34 +712,50 @@
          :headers {"Content-Type" "application/octet-stream"}
          :body (get-code-content environment code-id file-path)}))))
 
-(defn decode-catalog-post-body
-  [body]
+(def CatalogRequestV4
+  {(schema/required-key "certname") schema/Str
+   (schema/required-key "persistence") {(schema/required-key "facts") schema/Bool
+                                        (schema/required-key "catalog") schema/Bool}
+   (schema/required-key "environment") schema/Str
+   (schema/optional-key "trusted_facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/optional-key "facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/optional-key "job_id") schema/Str
+   (schema/optional-key "transaction_uuid") schema/Str
+   (schema/optional-key "options") {(schema/optional-key "capture_logs") schema/Bool
+                                    (schema/optional-key "prefer_requested_environment") schema/Bool}})
+
+(def CompileRequest
+  {(schema/required-key "certname") schema/Str
+   (schema/required-key "environment") schema/Str
+   (schema/required-key "code_ast") schema/Str
+   (schema/required-key "trusted_facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/required-key "facts") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/required-key "variables") {(schema/required-key "values") {schema/Str schema/Any}}
+   (schema/optional-key "job_id") schema/Str
+   (schema/optional-key "transaction_uuid") schema/Str
+   (schema/optional-key "options") {(schema/optional-key "capture_logs") schema/Bool
+                                    (schema/optional-key "log_level") (schema/enum "debug" "info" "warn")}})
+
+(defn valid-body
```

Comment:
more clear if named validate-body, makes the name obviously parallel to schema/validate

Reply:
oh i see you want the difference of returning the body after validating.... this is still a little weird but i'm not as sure

Reply:
`validated-body`?